### DISCUSSION TITLE
How can I optimize the rendering of cached data?

### DISCUSSION THREAD

**Discussion by ItMaga** [TanStack/query#4790] [Q&A]
When I use dynamic components in the SPA, such as tabs or pages, it takes longer to render with cached data while I move between them. This is expected behavior, but it leaves a bad mark on the user experience because there is a slight "freeze"

Stack:
 - Vue2
 - vue-query
 
So, what advice would you give in this situation? Maybe someone faced a similar problem.

I use such constructions as the following for data rendering:

```vue
<template>
  <QueryHandler :isLoading="isLoading" :isError="isError">
    <ul>
        <ListItem v-for="item of items" :key="item.id" :item="item" />
    </ul>
  </QueryHandler>
</template>

<script>
  export default {
    setup() {
      const { data: items, isLoading, isError } = useQuery(['items'], api.getItems);
  
      return { items, isLoading, isError };
    },
  }
</script>
```

**Comment by DamianOsipiuk** [comment]
Render optimizations are a different category of problems and not strictly related to vue-query.

If you experience freezes during large collections rendering, usually you want to look at:
- optimize your rendering tree, which might help in specific scenarios (reduce amount of items you are inserting/modifying into DOM, improve your computation algorithms by leveraging hashMaps for lookups etc.)
- introduce pagination/virtualization to reduce amount of data displayed at once

**Reply by louis-young** [reply]
Yeah, I agree with the sentiment above. You need to determine _what_ exactly is causing the performance issue before you can to reason about it and implement change.

Try profiling the application and using any developer tools available. Only once you understand what is causing the issue can you begin to investigate, implement and verify fixes 🙂

**Reply by ItMaga** [reply]
@DamianOsipiuk Thanks for advice, I'll certainly try it.
Yeah, I understand that this is not strictly related to vue-query. However, I thought that maybe this was not a rare problem.

**Reply by ItMaga** [reply]
@louis-young Thanks for feedback! 🙌
I think the main reason is that when there is cached data, the root component renders all the content before moving to a new page, tab, etc. And perhaps some kind of delayed rendering or delayed fetching of cached data could potentially solve this.
